### PR TITLE
ref: Fix typing of indexer service, add string batch abstraction [INGEST-1380]

### PR DESCRIPTION
**Pull request by untitaker**
Precursor to https://github.com/getsentry/sentry/pull/36531#issuecomment-1183805179=, make it so that one can test most of the message processing logic without having to provide an indexer.

Tests are not being refactored, let's just write new ones (such as for the rate limiter) in a way where it leverages this abstraction.

### CODE REVIEW

File: src/sentry/release_health/release_monitor/metrics.py [hunk 1]
```diff
@@ -162,10 +162,10 @@ def fetch_project_release_health_totals(
                         env_name = indexer.reverse_resolve(row[env_key])
                         release_name = indexer.reverse_resolve(row[release_key])
                         row_totals = totals[row["project_id"]].setdefault(
-                            env_name, {"total_sessions": 0, "releases": defaultdict(int)}
+                            env_name, {"total_sessions": 0, "releases": defaultdict(int)}  # type: ignore
```
Comment (untitaker) [inline src/sentry/release_health/release_monitor/metrics.py:165]: This code's typing is wrong, but I don't know how to fix it. Doing something like adding an assertion `assert env_name is not None` actually just breaks tests.

**Review by onewland — COMMENTED** [review]
I think this looks fine, is my read correct that there are absolutely no behavioral changes? Just splitting up a giant method into multiple methods?

**Comment by untitaker** [comment]
yea

**Review by onewland — APPROVED** [review]

**Review by fpacifici — APPROVED** [review]
thanks

**Pull request merged**